### PR TITLE
Allow "out" columns

### LINKED ISSUE #2074: Schema fails to compile when column is named `out`
### Versions

- **Rust:** 1.35.0
- **Diesel:** 1.4.2

## Problem Description

When using `print-schema` on a database with columns named "out", the resulting schema will fail to compile with many errors like the following:

```
error[E0530]: function parameters cannot shadow unit structs
   --> src\schema.rs:287:1
    |
287 |    table! {
    |   _^
    |  |_|
    | ||
288 | ||     execution_instance (id) {
289 | ||         id -> Binary,
290 | ||         time_executed -> Nullable<Datetime>,
...   ||
306 | ||     }
307 | || }
    | || ^
    | ||_|
    | |__cannot be named the same as a unit struct
    |    the unit struct `out` is defined here
    |
    = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)
```

## Checklist

- [x] I have already looked over the [issue tracker](https://github.com/diesel-rs/diesel/issues) for similar issues.
- [x] This issue can be reproduced on Rust's stable channel. (Your issue will be
  closed if this is not the case)

### PR DESCRIPTION
**Pull request by Diggsey**
Fixes #2074

### CODE REVIEW

File: diesel/src/macros/mod.rs [hunk 1]
```diff
@@ -23,10 +23,11 @@ macro_rules! __diesel_column {
             DB: $crate::backend::Backend,
             <$table as QuerySource>::FromClause: QueryFragment<DB>,
         {
-            fn walk_ast(&self, mut out: $crate::query_builder::AstPass<DB>) -> $crate::result::QueryResult<()> {
-                $table.from_clause().walk_ast(out.reborrow())?;
-                out.push_sql(".");
-                out.push_identifier($sql_name)
+            #[allow(non_snake_case)]
+            fn walk_ast(&self, mut diesel__out: $crate::query_builder::AstPass<DB>) -> $crate::result::QueryResult<()> {
```
Comment (sgrif) [inline diesel/src/macros/mod.rs]: In procedural macros we tend to just prefix with `__`. Do you think it's worth being consistent with that convention here?

File: diesel_cli/src/infer_schema_internals/inference.rs [hunk 2]
```diff
@@ -14,6 +14,12 @@ static RESERVED_NAMES: &[&str] = &[
     "type", "typeof", "unsafe", "unsized", "use", "virtual", "where", "while", "yield",
 ];
 
+fn is_reserved(name: &str) -> bool {
+    RESERVED_NAMES.contains(&name) || (
+        name.starts_with("diesel__") && !name.ends_with("_")
```
Comment (sgrif) [inline diesel_cli/src/infer_schema_internals/inference.rs]: I'm a little confused by the `ends_with` check. It seems unrelated?

File: diesel/src/macros/mod.rs (hunk 1 above)
Comment (Diggsey) [reply diesel/src/macros/mod.rs]: To me, a leading underscore is an indication that the variable is not necessarily used, as it will disable "unused variable" warnings in the compiler.

Also, it seems plausible that people could have columns that start with an underscore. It seems less plausible that they will have columns starting `diesel__`.

File: diesel_cli/src/infer_schema_internals/inference.rs (hunk 2 above)
Comment (Diggsey) [reply diesel_cli/src/infer_schema_internals/inference.rs]: It's not strictly necessary, it's just that we append an underscore as a way to make something "not reserved", so it seemed simpler and more consistent to just treat everything ending in an underscore as not a reserved word.

File: diesel/src/macros/mod.rs (hunk 1 above)
Comment (Diggsey) [reply diesel/src/macros/mod.rs]: (To clarify: reserving this namespace is a breaking change because it will change people's existing schema, and reserving `diesel__` seems less likely to cause problems than reserving `__`)

Comment (sgrif) [reply diesel/src/macros/mod.rs]: > Also, it seems plausible that people could have columns that start with an underscore.

Right, which is why we use double leading underscore

Comment (Diggsey) [reply diesel/src/macros/mod.rs]: I do not feel particularly strongly either way so have updated the PR.

**Comment by Diggsey** [comment]
These build failures seem unrelated to my changes.
